core/formatter.cpp

```diff
@@ -1724,16 +1733,233 @@ class FixIndentation {
+                LocationRange(),
+                fodder,
+                binds,
+                body
+            );
+            body = local;
+        }
 
-// TODO(dcunnin): Add pass to alphabeticize top level imports.
+        return body;
+    }
 
+    bool duplicatedVariables(const ImportElems &elems) {
+        std::vector<UString> idents;
+        for (const auto &elem: elems) {
+            idents.push_back(elem.bind->var->name);
+        }
+        return std::unique(idents.begin(), idents.end()) != idents.end();
+    }
+
+    /// Check if the import group ends after this local
+    bool groupBoundaryAfter(Local *local) {
+        Local *next = goodLocalOrNull(local->body);
+        if (!next) {
+            return true;
+        }
+
+        bool newlineReached = false;
+        for (const auto &fodderElem: open_fodder(next)) {
+            if (newlineReached || fodderElem.blanks > 0) {
+                return true;
+            }
+            if (fodderElem.kind != FodderElement::Kind::INTERSTITIAL) {
+                newlineReached = true;
+            }
+        }
+        return false;
+    }
+
+    void ensureCleanNewline(Fodder &fodder) {
+        if (fodder.size() == 0 || fodder.back().kind == FodderElement::Kind::INTERSTITIAL) {
+            fodder.push_back(FodderElement(FodderElement::Kind::LINE_END, 0, 0, {}));
+        }
+    }
+
+    AST *toplevelImport(Local *local, ImportElems &imports, const Fodder &groupOpenFodder) {
+        assert(isGoodLocal(local));
+
+        Fodder commentAfter, beforeNextFodder;
+        std::tie(commentAfter, beforeNextFodder) = splitOpenFodder(local->body);
+
+        ensureCleanNewline(commentAfter);
+
+        ImportElems newImports = extractImportElems(local, commentAfter);
+        imports.insert(imports.end(), newImports.begin(), newImports.end());
+
+        if (groupBoundaryAfter(local)) {
+            sortGroup(imports);
+
+            Fodder afterGroup = imports.back().afterFodder;
+            ensureCleanNewline(beforeNextFodder);
+            auto nextOpenFodder = concat_fodder(
+                afterGroup,
+                beforeNextFodder
+            );
+
+            // Process the code after the current group:
+            AST *bodyAfterGroup;
+            Local *next = goodLocalOrNull(local->body);
+            if (next) {
+                // Another group of imports
+                ImportElems nextImports;
+                bodyAfterGroup = toplevelImport(next, nextImports, nextOpenFodder);
+            } else {
+                // Something else
+                bodyAfterGroup = local->body;
+                open_fodder(bodyAfterGroup) = nextOpenFodder;
+            }
+
+            return buildGroupAST(imports, bodyAfterGroup, groupOpenFodder);
+        } else {
+            assert(beforeNextFodder.empty());
+            return toplevelImport(dynamic_cast<Local*>(local->body), imports, groupOpenFodder);
+        }
+    }
+
+    void file(AST *&body)
```

Comment:
Most (but not all) methods in this file have it on a new line, so I'll convert to that. Maybe I should try running clang-format over the whole code (not this change ofc) and maybe I can add it to automatic tests so that we have it always consistent.